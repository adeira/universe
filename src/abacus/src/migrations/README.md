```js
// TODO: move to a bit more declarative model for certain scenarios. For example, we should keep
// the latest JSON schemas for our collections and instead of migrating them we should just consider
// them to be the source of truth and modify the DB accordingly to the latest version without any
// migrations.
```

These migrations are written in Rust on purpose. It allows you to:

- do whatever you could do from Rust code (not limited to some YAML or AQL syntax)
- migrate not only structure but manipulate data as well
- write integration tests for these migrations

As a disadvantage they might be more verbose.
