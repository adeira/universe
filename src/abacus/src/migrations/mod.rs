#![allow(non_snake_case)]

use crate::arangodb::ConnectionPool;
use arangors::document::options::InsertOptions;
use arangors::ClientError;
use futures::future::BoxFuture;

mod utils;

mod v0__migrations_init;
mod v1__create_collection_users_sessions;
mod v2__create_document_user_anonymous;
mod v3__create_document_user_admin;
mod v4__create_collection_products;
mod v5__create_view_search_products;
mod v6__create_collection_tracking;
mod v7__create_collection_pos_checkouts;

#[derive(serde::Serialize, serde::Deserialize)]
struct MigrationRecord {
    _key: String,
}

/// All migrations should be written to be idempotent: not to fail when they are run repeatedly.
/// Please note: the migration engine currently doesn't care about the order or anything (it's quite naive).
pub async fn migrate(pool: &ConnectionPool) {
    let db = pool.db().await;

    // https://users.rust-lang.org/t/error-distinct-uses-of-impl-trait-result-in-different-opaque-types/46862/2
    // https://users.rust-lang.org/t/how-to-handle-a-vector-of-async-function-pointers/39804
    // https://users.rust-lang.org/t/how-to-store-async-functions-in-a-vector/51630
    type MigrationFunction = fn(
        &arangors::Database<uclient::reqwest::ReqwestClient>,
    ) -> BoxFuture<'_, Result<(), ClientError>>;

    // TODO: hide behind some macro (?)
    let migrations: Vec<(&str, MigrationFunction)> = vec![
        ("v0__migrations_init", |db| {
            Box::pin(v0__migrations_init::migrate(&db))
        }),
        ("v1__create_collection_users_sessions", |db| {
            Box::pin(v1__create_collection_users_sessions::migrate(&db))
        }),
        ("v2__create_document_user_anonymous", |db| {
            Box::pin(v2__create_document_user_anonymous::migrate(&db))
        }),
        ("v3__create_document_user_admin", |db| {
            Box::pin(v3__create_document_user_admin::migrate(&db))
        }),
        ("v4__create_collection_products", |db| {
            Box::pin(v4__create_collection_products::migrate(&db))
        }),
        ("v5__create_view_search_products", |db| {
            Box::pin(v5__create_view_search_products::migrate(&db))
        }),
        ("v6__create_collection_tracking", |db| {
            Box::pin(v6__create_collection_tracking::migrate(&db))
        }),
        ("v7__create_collection_pos_checkouts", |db| {
            Box::pin(v7__create_collection_pos_checkouts::migrate(&db))
        }),
    ];

    tracing::info!("Checking pending database migrations");
    for (migration_name, migration) in &migrations {
        if let Ok(migrations) = db.collection("migrations").await {
            if migrations
                .document::<MigrationRecord>(migration_name)
                .await
                .is_ok()
            {
                // the migration was already applied before - skip
                continue;
            }
        }

        let migration_result = migration(&db).await;
        match migration_result {
            Ok(_) => {
                // migration was successful - let's remember it so we don't run it again later
                if let Ok(migrations) = db.collection("migrations").await {
                    let result_future = migrations.create_document(
                        MigrationRecord {
                            _key: migration_name.to_string(),
                        },
                        InsertOptions::builder().return_new(true).build(),
                    );
                    match result_future.await {
                        Ok(_) => tracing::info!("{} ✅", migration_name),
                        Err(e) => {
                            tracing::error!("{} ❌", migration_name);
                            panic!("{:?}", e)
                        }
                    }
                }
            }
            Err(e) => {
                tracing::error!("{} ❌", migration_name);
                panic!("{:?}", e)
            }
        }
    }
    tracing::info!("Migrations OK");
}
