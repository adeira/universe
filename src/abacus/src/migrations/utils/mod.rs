use arangors::collection::options::{CreateOptions, CreateParameters};
use arangors::collection::CollectionType;
use arangors::document::options::InsertOptions;
use arangors::graph::Graph;
use arangors::index::Index;
use arangors::view::ViewOptions;
use serde::{Deserialize, Serialize};

pub(in crate::migrations) trait ArangoDocument {
    /// This key makes sure we don't create the same document twice by accident
    /// when running the migrations repeatedly.
    fn idempotency_key(&self) -> &str;
}

pub(in crate::migrations) async fn create_collection(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
    collection_name: &str,
    collection_type: &CollectionType,
    json_schema: &Option<serde_json::Value>,
) -> anyhow::Result<()> {
    match db.collection(collection_name).await {
        Ok(_) => {
            // collection already exists, nothing to do
            Ok(())
        }
        Err(_) => {
            // collection doesn't exist yet, let's create it
            let create_options = match json_schema {
                Some(json_schema) => CreateOptions::builder()
                    .name(collection_name)
                    .collection_type(*collection_type)
                    .schema(json_schema.to_owned())
                    .build(),
                None => CreateOptions::builder()
                    .name(collection_name)
                    .collection_type(*collection_type)
                    .build(),
            };
            let result = db
                .create_collection_with_options(create_options, CreateParameters::default())
                .await;

            match result {
                Ok(_) => Ok(()),
                Err(e) => anyhow::bail!(e),
            }
        }
    }
}

pub(in crate::migrations) async fn create_document<T>(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
    collection_name: &str,
    document: T,
) -> anyhow::Result<()>
where
    T: Serialize + for<'de> Deserialize<'de> + ArangoDocument,
{
    let collection = db.collection(&collection_name).await.unwrap(); // TODO
    match collection.document::<T>(document.idempotency_key()).await {
        Ok(_) => {
            // user already exists
            Ok(())
        }
        Err(_) => {
            // user doesn't exist yet, let's create it
            let result = collection
                .create_document(document, InsertOptions::builder().return_new(true).build())
                .await;

            match result {
                Ok(_) => Ok(()),
                Err(e) => anyhow::bail!(e),
            }
        }
    }
}

pub(in crate::migrations) async fn create_index(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
    collection_name: &str,
    index: &Index,
) -> anyhow::Result<()> {
    match db.create_index(collection_name, &index).await {
        Ok(_) => Ok(()),
        Err(e) => anyhow::bail!(e),
    }
}

pub(in crate::migrations) async fn create_graph(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
    graph: Graph,
) -> anyhow::Result<()> {
    match db.graph(&*graph.name).await {
        Ok(_) => {
            // graph already exists
            Ok(())
        }
        Err(_) => {
            // graph doesn't exist yet, let's create it
            match db.create_graph(graph, false).await {
                Ok(_) => Ok(()),
                Err(e) => anyhow::bail!(e),
            }
        }
    }
}

pub(in crate::migrations) async fn create_view(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
    view_name: &str,
    view: ViewOptions,
) -> anyhow::Result<()> {
    match db.view(&view_name).await {
        Ok(_) => {
            // view already exists
            Ok(())
        }
        Err(_) => {
            // view doesn't exist yet, let's create it
            match db.create_view(view).await {
                Ok(_) => Ok(()),
                Err(e) => anyhow::bail!(e),
            }
        }
    }
}
