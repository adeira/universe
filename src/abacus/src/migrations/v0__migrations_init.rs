use crate::migrations::utils::create_collection;
use arangors::collection::CollectionType;
use arangors::ClientError;

pub async fn migrate(
    db: &arangors::Database<arangors::client::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    let collection_name = "migrations";
    create_collection(&db, &collection_name, &CollectionType::Document, &None).await
}
