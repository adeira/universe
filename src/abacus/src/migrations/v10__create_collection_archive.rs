use crate::migrations::utils::create_collection;
use arangors::collection::CollectionType;

pub async fn migrate(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
) -> anyhow::Result<()> {
    create_collection(
        &db,
        "archive",
        &CollectionType::Document,
        &serde_json::from_str(std::include_str!("json_schemas/archive.json")).unwrap(),
    )
    .await
}
