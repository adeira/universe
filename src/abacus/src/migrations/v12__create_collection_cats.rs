use crate::arango::collection::CollectionType;
use crate::arango::DatabaseType;
use crate::migrations::utils::create_collection;

pub async fn migrate(db: &DatabaseType) -> anyhow::Result<()> {
    create_collection(db, "cats", &CollectionType::Document, &None).await
}
