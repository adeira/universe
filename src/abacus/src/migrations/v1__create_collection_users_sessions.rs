use crate::migrations::utils::{create_collection, create_graph, create_index};
use arangors::collection::CollectionType;
use arangors::graph::{EdgeDefinition, Graph};
use arangors::index::{Index, IndexSettings};
use arangors::ClientError;

pub async fn migrate(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    // 1. create `users` table
    create_collection(
        &db,
        "users",
        &CollectionType::Document,
        &serde_json::from_str(std::include_str!("json_schemas/users.json")).unwrap(),
    )
    .await?;

    // 2. create `sessions` table
    create_collection(
        &db,
        "sessions",
        &CollectionType::Document,
        &serde_json::from_str(
            r##"
            {
              "message": "",
              "level": "strict",
              "rule": {
                "type": "object",
                "additionalProperties": false,
                "required": ["last_access", "type"],
                "properties": {
                  "last_access": {
                    "type": "string"
                  },
                  "type": {
                    "type": "string",
                    "enum": ["mobile", "webapp"]
                  }
                }
              }
            }
            "##,
        )
        .unwrap(),
    )
    .await?;

    // 3. create edge between `users` and `sessions`
    create_collection(&db, "user_sessions", &CollectionType::Edge, &None).await?;

    // 4. create TTL index for the sessions
    create_index(
        &db,
        "sessions",
        &Index::builder()
            .name("last_access_mobile_ttl_index") // required in migrations for idempotency
            .fields(vec!["last_access".to_string()])
            .settings(IndexSettings::Ttl {
                // Mobile session token is long-lived (one year = 365 * 24 * 60 * 60)
                expire_after: 31_536_000,
            })
            .build(),
    )
    .await?;

    // 5. create sessions graph
    create_graph(
        &db,
        Graph::builder()
            .name(String::from("sessions"))
            .edge_definitions(vec![EdgeDefinition {
                collection: String::from("user_sessions"),
                from: vec![String::from("users")],
                to: vec![String::from("sessions")],
            }])
            .build(),
    )
    .await
}
