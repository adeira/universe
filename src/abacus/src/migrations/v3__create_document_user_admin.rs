use crate::migrations::utils::{create_document, ArangoDocument};
use arangors::ClientError;

#[derive(serde::Serialize, serde::Deserialize)]
struct User {
    _key: String,
    is_active: bool,
    google: UserGoogleClaims,
}

#[derive(serde::Serialize, serde::Deserialize)]
struct UserGoogleClaims {
    iss: String,
    sub: String,
    azp: String,
    aud: String,
    iat: usize,
    exp: usize,
    email: String,
    email_verified: bool,
    name: String,
    picture: String,
    given_name: String,
    family_name: String,
    locale: String,
}

impl ArangoDocument for User {
    fn idempotency_key(&self) -> &str {
        &self._key
    }
}

pub async fn migrate(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
) -> Result<(), ClientError> {
    let user_google_claims = UserGoogleClaims {
        iss: String::from("https://accounts.google.com"),
        sub: String::from("108269453578187886435"),
        azp: String::from("586578400209-5k61strd7i03b7pr2arn38o9qqghgfeq.apps.googleusercontent.com"),
        aud: String::from("586578400209-5k61strd7i03b7pr2arn38o9qqghgfeq.apps.googleusercontent.com"),
        iat: 1610381356,
        exp: 1610384956,
        email: String::from("mrtnzlml@gmail.com"),
        email_verified: true,
        name: String::from("Martin Zlámal"),
        picture: String::from("https://lh3.googleusercontent.com/a-/AOh14Gg_ZmWYHHMei_AQdckVaPqLikBkNYJONyyR0QqWfWs=s96-c"),
        given_name: String::from("Martin"),
        family_name: String::from("Zlámal"),
        locale: String::from("en")
    };

    let user = User {
        _key: String::from("2"),
        is_active: true,
        google: user_google_claims,
    };

    create_document(&db, "users", user).await
}
