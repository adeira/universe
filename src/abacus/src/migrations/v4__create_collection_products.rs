use crate::migrations::utils::{create_collection, create_document, ArangoDocument};
use arangors::collection::CollectionType;

#[derive(serde::Serialize, serde::Deserialize)]
struct ProductUnit {
    _key: String,
    en_US: String,
    es_MX: String,
}

impl ArangoDocument for ProductUnit {
    fn idempotency_key(&self) -> &str {
        &self._key
    }
}

pub async fn migrate(
    db: &arangors::Database<uclient::reqwest::ReqwestClient>,
) -> anyhow::Result<()> {
    create_collection(
        &db,
        "products",
        &CollectionType::Document,
        &serde_json::from_str(std::include_str!("json_schemas/products.json")).unwrap(),
    )
    .await?;

    create_collection(&db, "product_units", &CollectionType::Document, &None).await?;

    create_document(
        &db,
        "product_units",
        ProductUnit {
            _key: String::from("piece"),
            en_US: String::from("piece"),
            es_MX: String::from("pieza"),
        },
    )
    .await
}

#[cfg(test)]
mod tests {
    use serde_json::json;
    use valico::json_schema;

    #[test]
    fn products_json_schema_validation_test() {
        let arangodb_schema = serde_json::from_str::<serde_json::Value>(std::include_str!(
            "json_schemas/products.json"
        ))
        .unwrap();

        let mut scope = json_schema::Scope::new();
        let schema = scope
            .compile_and_return(arangodb_schema.get("rule").unwrap().clone(), false)
            .unwrap();

        // Valid input:
        assert!(schema
            .validate(&json!({
              "images": [],
              "unit_label": "TKTK",
              "is_published": false,
              "created": "TKTK",
              "visibility": ["POS", "ESHOP"],
              "categories": [],
              "updated": "TKTK",
              "price": {
                "unit_amount": 42,
                "unit_amount_currency": "MXN"
              },
              "translations": [
                {
                  "locale": "en_US",
                  "name": "TKTK",
                  "description": "TKTK"
                }
              ]
            }))
            .is_valid());

        // Invalid input (empty strings should not be allowed):
        insta::assert_json_snapshot!(schema.validate(&json!({
          "images": [],
          "unit_label": "",
          "is_published": false,
          "created": "",
          "visibility": [],
          "categories": [""],
          "updated": "",
          "price": {
            "unit_amount": 42,
            "unit_amount_currency": "MXN"
          },
          "translations": [
            {
              "locale": "en_US",
              "name": "",
              "description": ""
            },
            {
              "locale": "es_MX",
              "name": "",
              "description": ""
            }
          ]
        })));

        // missing fields should be captured:
        insta::assert_json_snapshot!(schema.validate(&json!({})));
    }
}
