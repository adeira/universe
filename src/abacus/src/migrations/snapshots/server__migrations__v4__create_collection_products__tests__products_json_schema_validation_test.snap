---
source: src/migrations/v4__create_collection_products.rs
expression: "schema.validate(&json!({\n                           \"images\" : [], \"unit_label\" : \"\", \"is_published\" :\n                           false, \"created\" : \"\", \"visibility\" : [], \"updated\"\n                           : \"\", \"price\" :\n                           {\n                               \"unit_amount\" : 42, \"unit_amount_currency\" :\n                               \"MXN\"\n                           }, \"translations\" :\n                           [{\n                                \"locale\" : \"en_US\", \"name\" : \"\", \"description\"\n                                : \"\"\n                            },\n                            {\n                                \"locale\" : \"es_MX\", \"name\" : \"\", \"description\"\n                                : \"\"\n                            }]\n                       }))"
---
{
  "errors": [
    {
      "code": "min_length",
      "path": "/created",
      "title": "MinLength condition is not met"
    },
    {
      "code": "min_length",
      "path": "/translations/0/description",
      "title": "MinLength condition is not met"
    },
    {
      "code": "min_length",
      "path": "/translations/0/name",
      "title": "MinLength condition is not met"
    },
    {
      "code": "min_length",
      "path": "/translations/1/description",
      "title": "MinLength condition is not met"
    },
    {
      "code": "min_length",
      "path": "/translations/1/name",
      "title": "MinLength condition is not met"
    },
    {
      "code": "min_length",
      "path": "/unit_label",
      "title": "MinLength condition is not met"
    },
    {
      "code": "min_length",
      "path": "/updated",
      "title": "MinLength condition is not met"
    }
  ],
  "missing": []
}
